docs(ngrx/signals): mention re-use of `withComputed/withMethods`

michael-small opened the pull request:
Issue: https://github.com/ngrx/platform/issues/4669

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue: [docs(@ngrx/signals): add a mention of restructuring withComputed or withMethods for re-use in those features](https://github.com/ngrx/platform/issues/4669)

Closes #4669

## What is the new behavior?

Documentation mentions how to re-use a computed or function across or within a `withComputed` or `withMethods`

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

michael-small commented:
@timdeschryver @rainerhahnekamp since you two coached me in the issue

I tweaked the example code comments a little.

I added the block towards the end of the "Core Concepts > Defining Store Methods" as this pertains to both `withMethods` and `withComputed` which by then are both explained.

markostanimirovic commented on projects/ngrx.io/content/guide/signals/signal-store/index.md:
```diff
@@ -285,6 +285,38 @@ export const BooksStore = signalStore(
+export const BooksStore = signalStore(
+  withState(initialState),
+  // 👇 Accessing previously defined state signals and properties.
+  withComputed(({ books, filter }) => ({
+    booksCount: computed(() => books().length),
+    sortDirection: computed(() => filter.order() === 'asc' ? 1 : -1),
+  })),
+  // 👇 (A) Also access previously defined computed properties (or functions).
+  withComputed(({ books, sortDirection }) => {
```
When I was writing this page, I assumed that readers are familiar with core JavaScript concepts, so they're aware that `() => ({})` is a sugar syntax for `() => { return {} }`.

I'm not a fan of using multiple `withComputed` calls. The native JS way should be preferred and recommended in my opinion:

```ts
withComputed({ books, filter }) => {
  const booksCount = computed(() => books().length);
  const sortDirection = computed(() => /* ... */);
  // ...
  
  return { /* export public computed signals */ };
}),
```

Additionally, it provides the ability not to export private computed signals if necessary.

michael-small replied:
> When I was writing this page, I assumed that readers are familiar with core JavaScript concepts, so they're aware that () => ({}) is a sugar syntax for () => { return {} }.

(sorry if this is overselling but I have thought on this a lot) In retrospect that made sense when I learned this after seeking it out, but I have seen people with this same problem frequently across different platforms. It has even been news to some GDEs who make signal store content. I can't speak for why the various other people didn't recognize this sugar syntax, but for me I just think going from a class based store to a function based one has a lot of little moments like this. And this point for people who wonder about this seems to solve a large hurdle with a small tip.

> I'm not a fan of using multiple withComputed calls. The native JS way should be preferred and recommended in my opinion:

I agree. That said, in the places this has come up, it feels like 2/3 of people are like "great, no need for another `withComputed`" and the other 1/3 are like "that's cool that the `const` method works good but I prefer multiple `withComputed`". So in my opinion, I feel like this `const` helper restructuring is worth a tip somewhere and that people who prefer multiple `withComputed` are just going to keep doing that lol.

michael-small replied:
What about having this PR as still adding this tip with respect to the native `const` approach, but dropping the repeated `withComputed/withMethods` reference?

@timdeschryver you suggested having them both, so I'm curious if my proposed compromise which tentatively drop the repeated `withComputed` sounds legit.

markostanimirovic replied:
If both options look good to Tim and Rainer, then I'm also fine with adding both of them to the documentation.

However, I'd move this from "Core Principles" to the "FAQ" page with a few changes:

- Separate examples for each case (A) and (B).
- Concise examples that are focused on the tip. For example, the `sortBooks` helper is not needed to show how computed signals/methods can be reused.

timdeschryver replied:
Hey, sorry if I mentioned something that wasn't recommended.
I'm following @markostanimirovic's lead here, let's move it to the FAQ page as was originally planned. I do think that it can be useful to be mentioned.

michael-small replied:
I like @rainerhahnekamp's take on that - showing option B but giving a passing mention of option A in text. That seems inline from how I am reading everyone's take on this - emphasizing option B, but having some mention of option A.

Additionally, move this to a FAQ point.

michael-small commented:
Yeah @markostanimirovic, I was curious about what I think can be agreed upon since 3 maintainers all had their own takes. But I didn't want to ping everyone when you guys were working hard on a release.

Thoughts?

> I like rainerhahnekamp's take on that - showing option B but giving a passing mention of option A in text. That seems inline from how I am reading everyone's take on this - emphasizing option B, but having some mention of option A.
> 
> Additionally, move this to an FAQ point.

So to the FAQ for sure, but the former point I was wondering about before fleshing out the next iteration of the tip

michael-small commented:
@markostanimirovic @timdeschryver @rainerhahnekamp 

I did the following after making some assumptions on what would work well between everyone's takes:

- Moved the tip to the FAQ
- With respects of "Option A" (two consecutive feature blocks)
    - Removed a showcase of "A" in code
    - Mentioned that "A" was possible in text, but took the opportunity to say why it is not encouraged
- Streamlined the example by removing `sortBooks`.
- Adjusted the example to be concisely about a `computed` that is returned but also a `computed` made from that one in the same scope
    - Perhaps it is a bit contrived, but I think it is more concise and easy to follow
    - I didn't want to stray too far from the type of example that was workshopped starting in the issue. However, if this example as it is now seems off then I can rework it to be something else
